File: apps/api/forms/Component.Form.Application/PageHandler/Default/DefaultPageHandler.cs
using Component.Form.Application.ComponentHandler;
using Component.Form.Application.Helpers;
using Component.Form.Application.Shared.Infrastructure;
using Component.Form.Application.UseCase.GetDataForPage.Model;
using Component.Form.Application.UseCase.ProcessForm.Model;
using Component.Form.Model;

namespace Component.Form.Application.PageHandler.Default;

public class DefaultPageHandler : IPageHandler
{
    private readonly IFormStore _formStore;
    private readonly IFormDataStore _formDataStore;
    private readonly IComponentHandlerFactory _componentHandlerFactory;
    
    public DefaultPageHandler(IFormStore formStore, IFormDataStore formDataStore, IComponentHandlerFactory componentHandlerFactory)
    {
        _formStore = formStore;
        _formDataStore = formDataStore;
        _componentHandlerFactory = componentHandlerFactory;
    }

    public bool IsFor(string type)
    {
        return String.IsNullOrEmpty(type) || type.Equals("default", StringComparison.OrdinalIgnoreCase);
    }

    public static string GetSafeType()
    {
        return SafeJsonHelper.GetSafeType(typeof(PageBase));
    }

    public async Task<GetDataForPageResponseModel> Get(PageBase page, dynamic data, string extraData)
    {        
        var parsedDataAsDictionary = (IDictionary<string, object>)data;

        var errors = new Dictionary<string, List<string>>();

        // Loop through questions and validate them
        foreach (var question in page.Components.Where(c => c.IsQuestionType))
        {
            string inputName = question.Name;
            IComponentHandler handler = _componentHandlerFactory.GetFor(question.Type);

            // as we're in the GET side, if the key does not exist its because it's a new page
            if (!parsedDataAsDictionary.ContainsKey(inputName))
            {
                continue;
            }

            List<string> validationResult = await handler.Validate(inputName, parsedDataAsDictionary, question.ValidationRules);
            if (validationResult.Count > 0)
            {
                errors.Add(inputName, validationResult);
            }
        }
  
        var forThisPage = new Dictionary<string, object>();

        foreach (var question in page.Components.Where(c => c.IsQuestionType))
        {
            string inputName = question.Name;

            if (parsedDataAsDictionary.ContainsKey(inputName))
            {
                forThisPage.Add(inputName, parsedDataAsDictionary[inputName]);
            }
        }

        return new GetDataForPageResponseModel
        {
            FormData = forThisPage,
            Errors = errors,
        };
    }

    public async Task<ProcessFormResponseModel> Process(PageBase page, dynamic existingData, Dictionary<string, string> formData)
    {
        var existingDataAsDictionary = (IDictionary<string, object>)existingData;

        var newData = GetData(page, formData, _componentHandlerFactory);

        foreach (var item in newData)
        {
            existingDataAsDictionary[item.Key] = item.Value;
        }

        var errors = new Dictionary<string, List<string>>();

        // Loop through questions and validate them
        foreach (var question in page.Components.Where(c => c.IsQuestionType))
        {
            string inputName = question.Name;
            IComponentHandler handler = _componentHandlerFactory.GetFor(question.Type);

            // Evaluate validation rules
            if (!question?.Optional ?? true)
            {
                var validationResult = await handler.Validate(inputName, existingData, question.ValidationRules);
                if (validationResult.Count > 0)
                {
                    errors.Add(inputName, validationResult);
                }
            }
        }

        // If there are validation errors, return to the same page
        if (errors.Any())
        {
            return new ProcessFormResponseModel
            {
                Data = existingData,
                NextPageId = page.PageId,
            };
        }

        var nextPageIdResult = await GetNextPageId(page, existingData, null);
        return new ProcessFormResponseModel
        {
            Data = existingData,
            NextPageId = nextPageIdResult.NextPageId,
        };
    }

    public async Task<ProcessChangeInFormResponseModel> ProcessChange(FormModel formModel, PageBase page, dynamic existingData, Dictionary<string, string> formData)
    {
        var existingDataAsDictionary = (IDictionary<string, object>)existingData;

        var newData = GetData(page, formData, _componentHandlerFactory);

        foreach (var item in newData)
        {
            existingDataAsDictionary[item.Key] = item.Value;
        }

        var errors = new Dictionary<string, List<string>>();

        // Loop through questions and validate them
        foreach (var question in page.Components.Where(c => c.IsQuestionType))
        {
            string inputName = question.Name;
            IComponentHandler handler = _componentHandlerFactory.GetFor(question.Type);

            // Evaluate validation rules
            if (!question?.Optional ?? true)
            {
                var validationResult = await handler.Validate(inputName, existingData, question.ValidationRules);
                if (validationResult.Count > 0)
                {
                    errors.Add(inputName, validationResult);
                }
            }
        }

        // If there are validation errors, return to the same page
        if (errors.Any())
        {
            return new ProcessChangeInFormResponseModel
            {
                Data = existingData,
                NextPageId = page.PageId,
                HasErrors = true
            };
        }
        
        return new ProcessChangeInFormResponseModel
        {
            Data = existingData
        };
    }

    public async Task<NextPageIdResult> GetNextPageId(PageBase page, dynamic data, string extraData, string endExtraDataCondition = "")
    {
        var conditionMetResult = await MeetsCondition(page, data);
        if (conditionMetResult.MetCondition)
        {
            return new NextPageIdResult
            {
                NextPageId = conditionMetResult.NextPageId,
                ExtraData = null
            };
        }

        return new NextPageIdResult
        {
            NextPageId = page.NextPageId,
            ExtraData = null
        };
    }

    private Dictionary<string, object> GetData(PageBase currentPage, Dictionary<string, string> newData, IComponentHandlerFactory _componentHandlerFactory) 
    {
        var data = new Dictionary<string, object>();

        foreach (var question in currentPage.Components.Where(c => c.IsQuestionType))
        {
            string inputName = question.Name;

            if (newData.ContainsKey(inputName))
            {
                IComponentHandler handler = _componentHandlerFactory.GetFor(question.Type);

                if (handler.GetDataType().Equals(SafeJsonHelper.GetSafeType(typeof(string)))) 
                {
                    data[inputName] = newData[inputName];
                }
                else 
                {
                    data[inputName] = FormHelper.ParseData(newData[inputName], _componentHandlerFactory);
                }
            }
        }

        return data;
    }

    private async Task<ConditionResult> MeetsCondition(PageBase currentPage, IDictionary<string, object> data)
    {
        if (currentPage.Conditions != null)
        {
            var nextPageId = "";
            bool metCondition = false;

            foreach (var condition in currentPage.Conditions)
            {
                if (await ExpressionHelper.EvaluateCondition(condition.Expression, data))
                {
                    nextPageId = condition.NextPageId;
                    metCondition = true;
                }
            }

            if (metCondition)
            {
                return new ConditionResult()
                {
                    MetCondition = true,
                    NextPageId = nextPageId
                };
            }
        }

        return new ConditionResult { MetCondition = false };
    }

    public async Task<WalkResult> WalkToNextInvalidOrUnfilledPage(FormModel formModel, PageBase currentPage, IDictionary<string, object> data, string extraData)
    {
        // if data does not exist, return the current page and stop
        // if data is invalid, return the current page and stop
        foreach (var question in currentPage.Components.Where(c => c.IsQuestionType))
        {
            string inputName = question.Name;
            if (!data.ContainsKey(inputName))
            {
                return new WalkResult
                {
                    Page = currentPage,
                    Stop = true
                };
            }

            IComponentHandler handler = _componentHandlerFactory.GetFor(question.Type);

            if (!question?.Optional ?? true)
            {
                var validationResult = await handler.Validate(inputName, data, question.ValidationRules);
                if (validationResult.Count > 0)
                {
                    return new WalkResult
                    {
                        Page = currentPage,
                        Stop = true
                    };
                }
            }
        }

        // if there are conditions check whether we have met them. If so, return the next conditional page..
        if (currentPage.Conditions != null)
        {
            foreach (var condition in currentPage.Conditions)
            {
                if (await ExpressionHelper.EvaluateCondition(condition.Expression, data))
                {
                    return new WalkResult
                    {
                        Page = formModel.Pages.FirstOrDefault(p => p.PageId == condition.NextPageId),
                        Stop = false
                    };
                }
            }
        }

        // go to tne next page and continue..
        return new WalkResult
        {
            Page = formModel.Pages.FirstOrDefault(p => p.PageId == currentPage.NextPageId),
            Stop = false
        };
    }

    public async Task<Dictionary<string, object>> GetSubmissionData(PageBase page, IDictionary<string, object> data)
    {
        var submissionData = new Dictionary<string, object>();

        foreach (var question in page.Components.Where(c => c.IsQuestionType))
        {
            string inputName = question.Name;
            if (data.ContainsKey(inputName))
            {
                submissionData.Add(inputName, data[inputName]);
            }
        }

        return submissionData;
    }
}

File: apps/api/forms-runner/src/utils/pageHandler/defaultPageHandler.ts
import { PageHandler } from '@/utils/pageHandler/interfaces';
import { ComponentHandlerFactory } from '@/utils/componentHandler/componentHandlerFactory';
import { meetsCondition } from '@/utils/expressionUtils';
import { getAllDataFromApplication } from '@/utils/applicationUtils';
import { Application } from '@model/formTypes';

export class DefaultPageHandler implements PageHandler {

    async Process(application: Application, pageId: string, data: { [key: string]: any; }): Promise<boolean> {
        const page = application.pages.find(p => p.pageId === pageId);
        if (!page) {
            throw new Error(`Page with ID ${pageId} not found.`);
        }

        let hasErrors = false;

        for (const component of page.components) {
        
            if (!component.type) {
                throw new Error(`Component type is undefined for component ID ${component.questionId}.`);
            }
        
            if (component.type === 'html') {
                // Skip HTML components
                continue;
            }

            const componentHandler = ComponentHandlerFactory.For(component.type);
            if (!componentHandler) {
                throw new Error(`No handler found for component type ${component.type}.`);
            }

            // update the component answer..
            if (component.name) {
                const convertedAnswer = componentHandler.Convert(component, data);
                console.log(`Converted answer for component ${component.name}:`, convertedAnswer);
                component.answer = convertedAnswer;
            } else {
                throw new Error(`Component name is undefined for component ID ${component.questionId}.`);
            }

            const validationResult = await componentHandler.Validate(component, getAllDataFromApplication(application));
            if (validationResult.length > 0) {
                hasErrors = true;
                component.errors = validationResult;
            }
            else {
                // ensure that errors are cleared if there are no validation errors
                component.errors = [];
            }
        }

        return hasErrors;
    }

    async GetNextPageId(application: Application, pageId: string): Promise<{ nextPageId: string | undefined, nextPageType: string | undefined, extraData: string | undefined }> {
        const page = application.pages.find(p => p.pageId === pageId);
        if (!page) {
            throw new Error(`Page with ID ${pageId} not found.`);
        }

        const { metCondition, nextPageId } = await meetsCondition(page, getAllDataFromApplication(application));
        if (metCondition) {
            const nextPage = application.pages.find(p => p.pageId === nextPageId);
            return { nextPageId, nextPageType: nextPage?.pageType, extraData: undefined };
        }

        const nextPage = application.pages.find(p => p.pageId === page.nextPageId);
        return { nextPageId: page.nextPageId, nextPageType: nextPage?.pageType, extraData: undefined };
    }

    async WalkToNextInvalidOrUnfilledPage(application: Application, currentPageId: string, extraData: string,): Promise<{ pageId: string, pageType: string, stop: boolean }> {

        const page = application.pages.find(p => p.pageId === currentPageId);
        if (!page) {
            throw new Error(`Page with ID ${currentPageId} not found.`);
        }

        for (const component of page.components) {
        
            if (!component.type) {
                throw new Error(`Component type is undefined for component ID ${component.questionId}.`);
            }
        
            if (component.type === 'html') {
                // Skip HTML components
                continue;
            }

            // check whether the answer is empty. If so, we need to stop here.
            if (component.answer === undefined || component.answer === null || component.answer === "") {
                return { pageId: currentPageId, pageType: page.pageType || "", stop: true };
            }
            
            const componentHandler = ComponentHandlerFactory.For(component.type);
            if (!componentHandler) {
                throw new Error(`No handler found for component type ${component.type}.`);
            }

            const validationResult = await componentHandler.Validate(component, getAllDataFromApplication(application));
            if (validationResult.length > 0) {
                return { pageId: currentPageId, pageType: page.pageType || "", stop: true };
            }
        }

        const nextPageResult = await this.GetNextPageId(application, currentPageId);
        return { pageId: nextPageResult.nextPageId || "", pageType: nextPageResult.nextPageType || "", stop: false };
    }
}